src/frontends/sql/decode_test.cc

```diff
@@ -202,9 +203,104 @@ const TextprotoDeathMessagePair kTextprotoDeathMessagePairs[] = {
      .death_message = "Required field id was not present in Expression."},
     {.textproto = R"(id: 1)",
      .death_message = "Required field expr_variant not set."},
+    {.textproto = R"(id: 1 merge_operation: { })",
+     .death_message = "Each MergeOperation is expected to have at least one "
+                      "non-control input."},
 };
 
 INSTANTIATE_TEST_SUITE_P(DecodeExprDeathTest, DecodeExprDeathTest,
                          ValuesIn(kTextprotoDeathMessagePairs));
 
+// We test the decoding of the child `Operation`s in other tests. We also
+// ensure that we can correctly make a `MergeOperation` from two vectors of
+// inputs in another test. Therefore, this is strictly about whether we can
+// decode the `MergeOperation` proto. This test accepts 4 parameters: the id,
+// the name, the number of direct inputs, and the number of control inputs.
+// We will generate the given number of `Literal` values on the input proto.
+class DecodeMergeOpTest
+    : public TestWithParam<
+          std::tuple<std::optional<absl::string_view>, uint64_t, uint64_t>> {
+ protected:
+  DecodeMergeOpTest() : ir_context_(), decoder_context_(ir_context_) {}
+
+  std::string GenerateLiteralListContentsProto(uint64_t len,
+                                               uint64_t id_start) const {
+    std::vector<std::string> literal_vec;
+    for (uint64_t id = id_start; id < id_start + len; ++id) {
+      literal_vec.push_back(
+          absl::StrFormat(R"({ id: %u literal: { literal_str: "lit" } })", id));
+    }
+    return absl::StrJoin(literal_vec, ",");
+  }
+
+  std::string GetTextproto() const {
+    auto &[name, direct_size, control_size] = GetParam();
+    std::string name_str =
+        (name.has_value()) ? absl::StrFormat(R"(name: "%s")", *name) : "";
+    // Start from 2 so that we can assign 1 to the top level MergeOperation.
+    std::string direct_input_list_contents =
+        GenerateLiteralListContentsProto(direct_size, 2);
+    std::string control_input_list_contents =
+        GenerateLiteralListContentsProto(control_size, 2 + direct_size);
+    return absl::StrFormat(
+        "id: 1 %s merge_operation: { inputs: [ %s ] control_inputs: [ %s ] }",
+        name_str, direct_input_list_contents, control_input_list_contents);
+  }
+
+  Value GetDecodedValue() {
+    Expression expr;
+    EXPECT_TRUE(
+        google::protobuf::TextFormat::ParseFromString(GetTextproto(), &expr))
+        << "Could not decode expr";
+    return DecodeExpression(expr, decoder_context_);
+  }
+
+  IRContext ir_context_;
+  DecoderContext decoder_context_;
+};
+
+TEST_P(DecodeMergeOpTest, DecodeMergeOpTest) {
+  auto &[opt_name, direct_size, control_size] = GetParam();
+
+  Expression expr;
+  EXPECT_TRUE(
+      google::protobuf::TextFormat::ParseFromString(GetTextproto(), &expr))
+      << "Could not decode expr";
+  Value value = DecodeExpression(expr, decoder_context_);
```

Comment:
Why are you not using `GetDecodedValue()`?

Reply:
It isn't defined/declared for `DecodeMergeOpTest`. I used the same test pattern for `Literal`s and `SourceTableColumn`s because they were both terminals having about the same form. But there are difficulties that pop up for inner nodes when trying to use that pattern; see the comment about why the number `1` is fixed in the test.